File: DFS/039.Combination-Sum/039.Combination-Sum.ts
function combinationSum(candidates: number[], target: number): number[][] {
  const ret: number[][] = [];
  const current: number[] = [];
  candidates.sort((a, b) => a - b);
  dfs(0);
  return ret;
  function dfs(index: number, sum = 0) {
    if (sum === target) {
      ret.push([...current]);
      return;
    }
    if (sum > target) {
      return;
    }
    for (let i = index; i < candidates.length; i++) {
      current.push(candidates[i]);
      dfs(i, sum + candidates[i]);
      current.pop();
    }
  }
}

File: DFS/039.Combination-Sum/039.Combination-Sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        candidates.sort()
        result = []
        current = []
        def dfs(index, sum):
            if sum > target:
                return
            
            if sum == target:
                result.append(current[:])

            for i in range(index, len(candidates)):
                current.append(candidates[i])
                dfs(i, sum + candidates[i])
                current.pop()

        dfs(0, 0)
        return result
